### ISSUE TITLE
IE 11 not supported

### ISSUE THREAD

**Issue by eps1lon** [sinonjs/sinon#1951]
**Describe the bug**
IE 11 environment is not supported with regards to syntax and methods. For example sinon has transitive dependencies to `has-flag` which is an arrow function and uses `String.prototype.startsWith`. Missing syntax support wouldn't be as bad since it can be included in babel-loader but adding polyfills is kind of rough since this would also mean that it wouldn't warn us if we use `startsWith`. Maybe there is also an option to only include polyfills for certain node_modules but both points can't really be attributed to IE 11 support.

**To Reproduce**
1. Build `sinon-esm.js`
2. search for `hasFlag` which is an arrow function and uses `startsWith`

**Expected behavior**
Support for IE 11 out of the box.

**Screenshots**
None

**Context (please complete the following information):**
We're trying run unit tests for our react components in IE 11.

* Library version: 7.1.1
* Environment: IE 11
* Example URL: none
* Other libraries you are using: babel, enzyme, karma, webpack

I guess enzyme will be another bottleneck which requires `airbnb-browser-shims` anyway and I couldn't find a solution to provide polyfills only to certain dependencies.

**Comment by fatso83** [comment]
We are running Sinon in IE11 in our own browser tests. IE11 does not support Ecmascript Modules, so it makes no sense to use the sinon-esm build.

Just use the Sinon.js bundle. Works everywhere.

**Comment by oliviertassinari** [comment]
@fatso83 The issue we are hitting with the `esm` version isn't with Ecmascript Modules, it's with the ES6 syntax, like arrow functions.

**Comment by fatso83** [comment]
https://saucelabs.com/beta/tests/475df00710394fc0bd6890fad7051580/commands#0

https://api.travis-ci.org/v3/job/456960446/log.txt

**Comment by fatso83** [comment]
@oliviertassinari, I get that, but if you are consuming an Ecmascript module you shouldn't expect it to be ES5 to begin with (as it's not). You *will* require some sort of transpilation step in your build pipeline if you build a bundle from ES modules, as this is not valid ES5, but it IS a ES module:
`export { _createFakeServerWithClock as createFakeServerWithClock };`

I suspect that your webpack config excludes node_modules from undergoing Babel transplantation, so you either need to selectively include Sinon or just import the non-esm bundle instead.

Chances are I just had the same issue with the Material UI React lib (what are the chances, Olivier!), so you could employ my fix: https://stackoverflow.com/q/53154986/200987

**Comment by fatso83** [comment]
But I do agree that it would be wise to make the most common thing as easy as possible, so if a simple config change can produce a valid ESM that will not cause problems in most builds, we will accept that PR.

**Comment by oliviertassinari** [comment]
> you shouldn't expect it to be ES5 to begin with

What should I expect it to be, can it have ES2018 syntax?

> Chances are I just had the same issue with the Material UI React lib (what are the chances, Olivier!)

I would love to know more about it. Right now the target with Material-UI is to have a 3 outputs. One for ES5, one for ESM (ES5 + es modules), one for evergreen browser (es) basically using all the official syntax, so ES2018.

**Comment by eps1lon** [comment]
I used the module build because it's easier to investigate since everything is in one place. This issue stays the same if I tell webpack to use the `main` entry and therefore commonJS modules.

The modules used are of no concern for the generated bundle.

**Comment by mantoni** [comment]
This discussion doesn't seem very fruitful. I'll try to explain the Sinon.JS teams view:

To begin with, the statement from [our latest API documentation page](https://sinonjs.org/releases/v7.1.1/) is what we feel bound to. It sais

> Sinon v7.1.1 is written as ES5.1 and __requires no transpiler or polyfills__ to run in the runtimes listed below.

The listed runtimes are

> - Firefox 45
> - Chrome 48
> - __Internet Explorer 11__
> - Edge 14
> - Safari 9
> - Node.js LTS versions

Abviously, these statements are made reagarding the standard Sinon.JS distribution, as it is downloadable from our web page and distributed via npm. The links provided above show the successful runs of our test suite on IE 11, among other browsers.

We provide an ESM bundle for environments where ESM is supported, to make it possible to consume Sinon in these environments. ESM is not supported by IE 11.

Now, what you're asking for is to make an ESM version of Sinon for a target environment that doesn't support ESM. This is out of the scope of this project.

What @fatso83 was trying to explain is, that if you have a need like this, you're most likely transpiling your code anyway, since IE 11 does not support ESM out of the box. We're therefore asking you to take the regular Sinon build, which _does_ support IE 11, and wrap it in an ESM bundle yourself, just as you're most likely doing it with other resources.

We will not re-open this, as it's not an issue with Sinon. I hope you understand.

**Comment by eps1lon** [comment]
@mantoni Please read my comment again and consider that `module` and `main` are only concerned about the module system not other syntax or language features.

**Comment by mroderick** [comment]
@eps1lon @oliviertassinari I am not against the ESM version of Sinon being ESM + ES5.1, if that solves problems for people. Would you care to submit a pull request to scratch your itch?

**Comment by eps1lon** [comment]
Just to be perfectly clear the commonJS version isn't ES5.1 either which should be showcased with https://github.com/eps1lon/sinon-ie11

The issue is that you have a transitive dependency to the `has-flag` package via the `main` entry which uses fat arrow syntax as well as String.prototype.startsWith.
```bash
$ npm ls has-flag
sinon-ie-11@1.0.0 C:\Development\src\js\sinon-ie-11
`-- sinon@7.1.1
  `-- supports-color@5.5.0
    `-- has-flag@3.0.0
```

This is fixable by setting mainFields to `['browser', 'main', 'module']` but that's not what I usually have to tell webpack if I'm building for the browser and since webpack defaults to `['browser', 'module', 'main']` I would expect libraries to consider this as well.

I'm going to check if I can fix your esm build with a similar approach but I have to investigate if rollup allows mainFields configuration.

**Comment by mantoni** [comment]
The CommonJS version is ES5.1. You can search through the `pkg/sinon.js` sources and you will neither find arrow functions nor any use of `startsWith`. That is because the transitive dependency you're referring to is not bundled because `supports-color` has a [a browser version](https://github.com/chalk/supports-color/blob/master/browser.js) which simply states that color is not supported.

Again, if you're using Sinon in a browser, use the browser package. If you want the browser package within an ESM, you're invited to send a PR to produce a new ESM + ES5.1 bundle for Sinon. The current bundle does not need fixing. It's not broken.

**Comment by eps1lon** [comment]
@mantoni  I wish I never mentioned the esm build. It has nothing to do with the issue. It's however troubling that you only consider your code as part of the package and not dependencies. I hope you don't have a similar attitude towards security issues in dependencies.

**Comment by mantoni** [comment]
We do consider dependencies as part of our library. We also take security very seriously and watch security alerts on the repository and npm audit closely.

Let me try again, without using the word esm: I understand that you're loading the Sinon sources into IE 11. You're obviously not loading the browser version, because the browser version does not have the issues you're referring to. You should be loading the browser version if you want to use Sinon in a browser. Currently, only the `pkg/sinon.js` file is built to run in browsers. Apparently you're trying to use Sinon in an unsupported way. Since this is open source, you're invited to make it work by sending a pull request.

**Comment by fatso83** [comment]
People in this thread might be interested in knowing that @Feiyang1 implemented a fix for this by adding a `browser` field in `package.json`. See https://github.com/sinonjs/sinon/pull/2060#issuecomment-511099707

There should be no additional configuration needed for Webpack/Babel builds after this has shipped.
